core/src/main/java/com/linecorp/armeria/client/retry/FibonacciBackoff.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.retry;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.math.LongMath.saturatedAdd;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.base.MoreObjects;
+
+final class FibonacciBackoff extends AbstractBackoff {
+    private final long initialDelayMillis;
+    private final long maxDelayMillis;
+
+    private final long[] precomputedDelays;
+
+    FibonacciBackoff(long initialDelayMillis, long maxDelayMillis) {
+        checkArgument(initialDelayMillis >= 0,
+                      "initialDelayMillis: %s (expected: >= 0)", initialDelayMillis);
+        checkArgument(initialDelayMillis <= maxDelayMillis, "maxDelayMillis: %s (expected: >= %s)",
+                      maxDelayMillis, initialDelayMillis);
+        this.initialDelayMillis = initialDelayMillis;
+        this.maxDelayMillis = maxDelayMillis;
+
+        final List<Long> precomputed = new ArrayList<>();
```

Comment:
How about precomputing only when there will be many values, like @anuraaga did in his pull request? https://github.com/line/armeria/pull/1983/files#diff-2d744bfdcc8222dbb89cabef4de01117R45-R60

Reply:
I thought of this well. The problem I saw with the Fibonacci sequence is that you already calculate the whole sequence up until that point to find out if it becomes big. For 30, however, this can be estimated by using the same check as we do for like for the exponential one `Math.pow(1.6..., 29)`, but I think in practice I would set the bar way lower when it is worth performing the precalculation for Fibonacci.

Reply:
Reading through the PR from @anuraaga again, I assume the purpose is the opposite. It ensures that the collection is not becoming too big. Having `maxDelay` = multiplier*initialDelay` would also perform the precalculation with a resulting Collection of two elements, I think.

In my Implementation, I kick off an iterative calculation of the 31st, 32nd... etc. delay then, capping the Collection size at 30 still. I directly make use of predeceding values (29th and 30th), while the exponential solution is not using precalculation at that point.

Reply:
Agree there's no reason to make the precomputation conditional and I removed it from my PR. I was thinking to optimize the case where there are few possible values of delay, but realize that it doesn't hurt to optimize the case when there are lots of values too.